solc_select/solc_select.py

```diff
@@ -1,30 +1,139 @@
 import argparse
+import contextlib
 import hashlib
 import json
-from zipfile import ZipFile
 import os
-import shutil
+import platform
 import re
+import shutil
+import subprocess
 import sys
 import urllib.request
 from pathlib import Path
-from packaging.version import Version
+from urllib.parse import urlparse
+from zipfile import ZipFile
+
 from Crypto.Hash import keccak
+from packaging.version import Version
+
 from .constants import (
-    LINUX_AMD64,
-    MACOSX_AMD64,
-    WINDOWS_AMD64,
-    EARLIEST_RELEASE,
-    SOLC_SELECT_DIR,
     ARTIFACTS_DIR,
     CRYTIC_SOLC_ARTIFACTS,
     CRYTIC_SOLC_JSON,
+    EARLIEST_RELEASE,
+    LINUX_AMD64,
+    MACOSX_AMD64,
+    SOLC_SELECT_DIR,
+    WINDOWS_AMD64,
 )
 from .utils import mac_binary_is_universal, mac_can_run_intel_binaries
 
 Path.mkdir(ARTIFACTS_DIR, parents=True, exist_ok=True)
 
 
+def get_arch() -> str:
+    """Get the current system architecture."""
+    machine = platform.machine().lower()
+    if machine in ["x86_64", "amd64"]:
+        return "amd64"
+    elif machine in ["aarch64", "arm64"]:
+        return "arm64"
+    elif machine in ["i386", "i686"]:
+        return "386"
+    return machine
+
+
+def check_emulation_available() -> bool:
+    """Check if x86_64 emulation is available."""
+    if get_arch() != "arm64":
+        return False
+
+    # On macOS, check for Rosetta 2
+    if sys.platform == "darwin":
+        # Check if we can run x86_64 binaries
+        try:
+            result = subprocess.run(["arch", "-x86_64", "true"], capture_output=True, check=False)
+            return result.returncode == 0
+        except (FileNotFoundError, OSError):
+            return False
```

Comment:
We have a function that does this already, `mac_can_run_intel_binaries()`

Reply:
Perfect! Replaced the custom implementation with the existing `mac_can_run_intel_binaries()` function. Much cleaner and avoids code duplication.